src/java/org/apache/cassandra/db/compaction/unified/AdaptiveController.java

```diff
@@ -0,0 +1,330 @@
+
+        if (intervalSec <= 0)
+            throw new IllegalArgumentException(String.format("Invalid configuration for interval, it should be positive: %d", intervalSec));
+
+        if (gain <= 0 || gain > 1)
+            throw new IllegalArgumentException(String.format("Invalid configuration for gain, it should be within (0,1]: %f", gain));
+
+        if (minCost <= 0)
+            throw new IllegalArgumentException(String.format("Invalid configuration for minCost, it should be positive: %d", minCost));
+
+        return new AdaptiveController(new SystemTimeSource(), env, W, survivalFactor, dataSetSizeMB, numShards, minSstableSizeMB, flushSizeOverrideMB, maxSpaceOverhead, maxSSTablesToCompact, expiredSSTableCheckFrequency, ignoreOverlapsInExpirationCheck, intervalSec, minW, maxW, gain, minCost);
+    }
+
+    public static Map<String, String> validateOptions(Map<String, String> options) throws ConfigurationException
+    {
+        String s;
+        s = options.remove(STARTING_W);
+        if (s != null)
+            Integer.parseInt(s);
+        options.remove(MIN_W);
+        if (s != null)
+            Integer.parseInt(s);
+        options.remove(MAX_W);
+        if (s != null)
+            Integer.parseInt(s);
+        options.remove(INTERVAL_SEC);
+        if (s != null)
+            Integer.parseInt(s);
+        options.remove(GAIN);
+        if (s != null)
+            Double.parseDouble(s);
+        options.remove(MIN_COST);
+        if (s != null)
+            Integer.parseInt(s);
+        return options;
+    }
+
+    @Override
+    void startup(UnifiedCompactionStrategy strategy, CostsCalculator calculator)
+    {
+        super.startup(strategy, calculator);
+        this.lastChecked = timeSource.nanoTime();
+    }
+
+    @Override
+    public int getW(int index)
+    {
+        return W;
+    }
+
+    @Override
+    public double getSurvivalFactor()
+    {
+        return survivalFactor;
+    }
+
+    @Override
+    @Nullable
+    public CostsCalculator getCalculator()
+    {
+        return calculator;
+    }
+
+    public int getInterval()
+    {
+        return intervalSec;
+    }
+
+    public int getMinW()
+    {
+        return minW;
+    }
+
+    public int getMaxW()
+    {
+        return maxW;
+    }
+
+    public double getGain()
+    {
+        return gain;
+    }
+
+    public int getMinCost()
+    {
+        return minCost;
+    }
+
+    @Override
+    public void onStrategyBackgroundTaskRequest()
+    {
+        if (!isRunning())
+            return;
+
+        long now = timeSource.nanoTime();
+        if (now - lastChecked < TimeUnit.SECONDS.toNanos(intervalSec))
+            return;
+
+        try
+        {
+            maybeUpdate(now);
+        }
+        finally
+        {
+            lastChecked = now;
+        }
+    }
+
+    private void maybeUpdate(long now)
+    {
+        // Each shard will act independently and so our target size of the same as the shard size
```

Comment:
"size of the same" -> size is the same

Reply:
Fixed in ca70af6057a9e32fbf19f61b016825788a373431.